### PR TITLE
report error when subquery in case-when returns empty set

### LINKED ISSUE #3483: report error when case when clause  has empty condition
if the subquery of case when clause in when condition and it returns empty set,  then the whole query return empty set,
this is because doris rewrite subquery in case when to cross join, if a subqery return empty the cross join return empty. 
e.g. 
```
select case when (select k1 sum from t1 where  k1 <0) > t2. then t2.k1 else 0 end a from t2;
```
this query will rewrite to 
```
select case when tt.k1 > t2. then t2.k1 else 0 end a from t2, (select k1 sum from t1 where  k1 <0 ) as tt;
```
if `select k1 sum from t1 where  k1 <0` is an empty set, then the query is empty, but in mysql this returns `else 0` branch as the resulet.
in doris we can not handle subquery very well, so shoud report an error when subquery is empty
step 1:
enhance assert row num node to support more asset conditions
step 2:
apply assert rule to subquery in case when clause

### PR DESCRIPTION
**Pull request by yangzhg**
fix #3483 step2, apply assert rule to subquery in case-when clause

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/CaseExpr.java [hunk 1]
```diff
@@ -186,6 +189,9 @@ public void analyzeImpl(Analyzer analyzer) throws AnalysisException {
             // Determine maximum compatible type of the then exprs seen so far.
             // We will add casts to them at the very end.
             Expr thenExpr = children.get(i + 1);
+            if (!thenExpr.getType().isScalarType()) {
+                throw new AnalysisException("subquery in case-when must return scala type");
+            }
```
Comment (wutiangan) [inline fe/src/main/java/org/apache/doris/analysis/CaseExpr.java:194]: why do you only thow exception where expr is ‘when expr’ and ‘then expr’？
Whether to throw an exception when ‘case expr’ and ‘else expr’ is not scarlar type？

File: fe/src/main/java/org/apache/doris/analysis/SelectStmt.java [hunk 2]
```diff
@@ -1391,9 +1391,8 @@ private void rewriteSelectList(ExprRewriter rewriter) throws AnalysisException {
             CaseExpr caseExpr = (CaseExpr) item.getExpr();
 
             int childIdx = 0;
-            if (caseExpr.hasCaseExpr()
-                    && caseExpr.getChild(childIdx++).contains(Predicates.instanceOf(Subquery.class))) {
-                throw new AnalysisException("Only support subquery in binary predicate in case statement.");
+            if (caseExpr.hasCaseExpr()) {
+                childIdx++;
             }
             while (childIdx + 2 <= caseExpr.getChildren().size()) {
```
Comment (wutiangan) [inline fe/src/main/java/org/apache/doris/analysis/SelectStmt.java:1397]: why whenExpr must be BinaryPredicate? can we support query “select case (select sex from table limit 1) when ''1” then 'male' else 'female'？

Comment (wutiangan) [inline fe/src/main/java/org/apache/doris/analysis/SelectStmt.java:1397]: maybe in 1397 line code , "<=" need to modify "<".
for example: case when expr0 then expr1 else expr2.   expr2 will be consider 'when expr' in In the following code

File: fe/src/main/java/org/apache/doris/analysis/CaseExpr.java (hunk 1 above)
Comment (yangzhg) [reply fe/src/main/java/org/apache/doris/analysis/CaseExpr.java:194]: if `when exp`or `case expr` has subquery and subquer returns non-scalar type, the error will be thrown by `analyzer.getCompatibleType` and the message is  weird, too difficult to understand, so i throw a message here

File: fe/src/main/java/org/apache/doris/analysis/SelectStmt.java (hunk 2 above)
Comment (yangzhg) [reply fe/src/main/java/org/apache/doris/analysis/SelectStmt.java:1397]: this is the previous code

Comment (yangzhg) [reply fe/src/main/java/org/apache/doris/analysis/SelectStmt.java:1397]: after the  loop  the `childIdx` is 2  `childIdx + 2` is 4 not <= `caseExpr.getChildren().size()` (3) so expr2 will not be consider 'when expr' in In the following code

**Review by EmmyMiao87 — DISMISSED** [review]
LGTM

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**